Windows tilde expansion followup

rkitover opened the pull request:
@justinmk @zeertzjq as promised, here's the stuff left over from our discussion in #28515.

rkitover committed:
refactor(main): os_get_homedir() for ~ expansion

Followup on fdeb01cd7 (feat(main): expand file ~\ or ~/ prefix on
Windows (#28515), 2024-05-25).

Use the os_get_homedir() accessor introduced in ddf6a7907 (feat(os/env):
add os_get_homedir() accessor, 2024-06-17) for ~/ expansion on Windows
to make the code more concise.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

zeertzjq commented on src/nvim/os/env.c:
```diff
@@ -412,6 +412,15 @@ void os_get_hostname(char *hostname, size_t size)
 static char *homedir = NULL;
 static char *os_homedir(void);
 
+const char *os_get_homedir(void)
```
Add a function about the difference of this function with `os_homedir()`?

zeertzjq replied:
Or maybe one of these functions should be renamed, as it's confusing to have both `os_homedir()` and `os_get_homedir()`.

rkitover replied:
You mean describe the differences in the commit message?

zeertzjq replied:
Sorry, I mean add a comment above this function.

justinmk replied:
does it make sense to rename the existing `os_homedir` to `uv_homedir` (and keep it private/static), then name this new function `os_homedir` ? And the docstring on the new function can mention:

    /// Gets the "real", resolved homedir. See comment on `homedir`.

zeertzjq replied:
`uv_` prefix is for libuv functions. This is a wrapper around a libuv function, so I think it should at least be `os_uv_homedir`.

rkitover replied:
This good?

zeertzjq replied:
I think you should reference the comment above the `homedir` variable.
> ```
> /// Gets the "real", resolved user home directory. See comment on `homedir`.
> ```

justinmk commented on test/functional/core/main_spec.lua:
```diff
@@ -194,3 +194,41 @@ describe('command-line option', function()
+
+if is_os('win') then
+  describe('Windows features', function()
+    it('expands ~/', function()
```
this it() can live in the describe() block starting on line 17. Else this is missing before_all(...).

```suggestion
    it('expands ~/ on Windows', function()
```

justinmk commented on test/functional/core/main_spec.lua:
```diff
@@ -194,3 +194,41 @@ describe('command-line option', function()
+      os.remove(fname)
+    end)
+
+    it('expands ~\\', function()
```
test looks very redundant, can the code be deduplicated

rkitover commented:
How's this?

rkitover committed:
refactor(os/env): s/os_homedir()/os_uv_homedir()/

Rename the static os_homedir() to os_uv_homedir() to emphasize that it
is a wrapper around a libuv function.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover committed:
feat(os/env): add os_get_homedir() accessor

Followup on fdeb01cd7 (feat(main): expand file ~\ or ~/ prefix on
Windows (#28515), 2024-05-25).

Add the function os_get_homedir() to os/env.c to return the cached
homedir value as a const. Must be called after homedir is initialized or
it fails.

The difference between this function and the static os_uv_homedir() is
that the latter gets the homedir from libuv and is used to initialize
homedir in init_homedir(), while os_get_homedir() just returns homedir
as a const if it's initialized and is public.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover commented:
Good to go now?

zeertzjq commented on test/functional/core/main_spec.lua:
```diff
@@ -193,4 +193,25 @@ describe('command-line option', function()
+    for _, prefix in ipairs({ '~/', '~\\' }) do
+      it('expands ' .. prefix .. ' on Windows', function()
+        local fname = os.getenv('USERPROFILE') .. '\\nvim_test.txt'
+        os.remove(fname)
+        write_file(fname, 'some text')
+        eq(
+          'some text',
+          fn.system({
+            nvim_prog_abs(),
+            '-es',
+            '+%print',
+            '+q',
+            prefix .. 'nvim_test.txt',
+          }):gsub('\n', '')
+        )
+        os.remove(fname)
```
I don't think it's necessary to use `os.remove()` before writing to the file. However, it's better to use `finally()`, so that the file is deleted even if the test fails.
```suggestion
        finally(function()
          os.remove(fname)
        end)
        write_file(fname, 'some text')
        eq(
          'some text',
          fn.system({
            nvim_prog_abs(),
            '-es',
            '+%print',
            '+q',
            prefix .. 'nvim_test.txt',
          }):gsub('\n', '')
        )
```

rkitover committed:
test(main): add ~/, ~\ expansion tests for Windows

Followup on fdeb01cd7 (feat(main): expand file ~\ or ~/ prefix on
Windows (#28515), 2024-05-25).

Add a Windows section to main_spec.lua with tests for expanding ~/ and
~\ prefixes for files passed in on the command-line.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover commented:
Fixed.

zeertzjq commented:
Because 0.11 isn't released yet

rkitover commented:
Sorry, I forgot how the release branches work.

I fixed the merge conflict [here](https://github.com/neovim/neovim/pull/31254).
